### PR TITLE
public_key: CA extended key usage check

### LINKED ISSUE #8806: Erlang Critical EKU on ICA certificate: Fatal - Unsupported Certificate
**Describe the bug**
When x509 Extended Key Usage (EKU) is marked as critical on a Intermediate Certificate Authority (ICA) certificate within a cert chain Erlang OTP errors out with.
```
{error,{tls_alert,{unsupported_certificate,"TLS client: In state wait_cert_cr at ssl_handshake.erl:2117 generated CLIENT ALERT: Fatal - Unsupported Certificate\n"}}}
```

OpenSSL, Postgres, Redis, JVM, Chrome, GoLang stdlib, and the .NET framework have no issues with these same certificate chains.

**To Reproduce**

1 Setup
Download attached zip file.  Extract it and navigate to the resulting directory in bash shell.

2 Create Root CA
```bash
openssl genrsa -out root.key 3072
openssl req -config root.cnf -key root.key -new -x509 -days 7300 -sha256 -extensions v3_ca -out root.crt -subj "/C=US/ST=California/L=Temecula/O=ACME Inc/OU=Labs/CN=Root CA/emailAddress=test@testing.io"
```
3 Good For Erlang
Only extension marked as critical in the cert chain is `Basic Constraints` as that is necessary for valid CA certificates with x509 strict validation.
3.1 Create Intermediate CA
```bash
openssl genrsa -out good_ica.key 3072
openssl req -config good_ica.cnf -key good_ica.key -new -sha256 -out good_ica.csr -subj "/C=US/ST=California/L=Temecula/O=ACME Inc/OU=GOOD/CN=Good ICA/emailAddress=good@testing.io"
openssl x509 -req -extfile good_ica.cnf -extensions v3_intermediate_ca -CA root.crt -CAkey root.key -CAcreateserial -days 3650 -in good_ica.csr -out good_ica.crt
```
3.2 Create Cert
```bash
openssl genrsa -out good_test.key 4096
openssl req -config good_ica.cnf -key good_test.key -new -sha256 -out good_test.csr -subj "/C=US/ST=California/L=Temecula/O=ACME Inc/OU=GOOD_ICA_IN_USE/CN=GOOD Leaf Cert"
openssl x509 -req -extfile good_ica.cnf -extensions server_cert -CA good_ica.crt -CAkey good_ica.key -CAcreateserial -days 3650 -in good_test.csr -out good_test.crt
```
3.3 Create CA Chain
```bash
cat good_ica.crt root.crt > good_ca-chain.crt
```

3.4 Run TLS Test Server
```bash
openssl s_server -accept 8443 -cert good_test.crt -key good_test.key -CAfile good_ca-chain.crt
```

3.5 Test TLS with OpenSSL
```bash
openssl s_client -connect 127.0.0.1:8443  -showcerts -CAfile good_ca-chain.crt -verify_return_error -x509_strict
```

Happy Dance - everything looks good
```openssl
---
Server certificate
subject=C=US, ST=California, L=Temecula, O=ACME Inc, OU=GOOD_ICA_IN_USE, CN=GOOD Leaf Cert
issuer=C=US, ST=California, L=Temecula, O=ACME Inc, OU=GOOD, CN=Good ICA, emailAddress=good@testing.io
---
No client certificate CA names sent
Peer signing digest: SHA256
Peer signature type: RSA-PSS
Server Temp Key: X25519, 253 bits
---
SSL handshake has read 5129 bytes and written 379 bytes
Verification: OK
---
New, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384
Server public key is 4096 bit
This TLS version forbids renegotiation.
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
```
3.6 Test with Erlang OTP 26 (in Erlang shell)
Command:
```erlang
ssl:start(), ssl:connect("127.0.0.1", 8443, [{verify, verify_peer}, {versions, ['tlsv1.3']}, {cacertfile, "good_ca-chain.crt"}, {server_name_indication, disable}]).
```

Full Result (Happy Dance):
```erlang
Erlang/OTP 26 [erts-14.2.5] [source] [64-bit] [smp:10:10] [ds:10:10:10] [async-threads:1] [jit] [dtrace]

Eshell V14.2.5 (press Ctrl+G to abort, type help(). for help)
1> ssl:start(), ssl:connect("127.0.0.1", 8443, [{verify, verify_peer}, {versions, ['tlsv1.3']}, {cacertfile, "bad_ca-chain.crt"}, {server_name_indication, disable}]).
{ok,{sslsocket,{gen_tcp,#Port<0.4>,tls_connection,undefined},
               [<0.120.0>,<0.119.0>]}}
```

4 Bad For Erlang
Critical is specified for Basic Constraints and EKU extensions for this test.  Only change in the ICA cert extensions from #3 is that EKU is now marked as critical.
4.1 Create Intermediate CA
```bash
openssl genrsa -out bad_ica.key 3072
openssl req -config bad_ica.cnf -key bad_ica.key -new -sha256 -out bad_ica.csr -subj "/C=US/ST=California/L=Temecula/O=ACME Inc/OU=BAD/CN=BAD ICA/emailAddress=bad@testing.io"
openssl x509 -req -extfile bad_ica.cnf -extensions v3_intermediate_ca -CA root.crt -CAkey root.key -CAcreateserial -days 3650 -in bad_ica.csr -out bad_ica.crt
```
4.2 Create Cert
```bash
openssl genrsa -out bad_test.key 4096
openssl req -config bad_ica.cnf -key bad_test.key -new -sha256 -out bad_test.csr -subj "/C=US/ST=California/L=Temecula/O=ACME Inc/OU=BAD_ICA_IN_USE/CN=BAD Leaf Cert"
openssl x509 -req -extfile bad_ica.cnf -extensions server_cert -CA bad_ica.crt -CAkey bad_ica.key -CAcreateserial -days 3650 -in bad_test.csr -out bad_test.crt
```
4.3 Create CA Chain
```bash
cat bad_ica.crt root.crt > bad_ca-chain.crt
```

4.4 Run TLS Test Server
```bash
openssl s_server -accept 8443 -cert bad_test.crt -key bad_test.key -CAfile bad_ca-chain.crt
```

4.5 Test TLS with OpenSSL
```bash
openssl s_client -connect 127.0.0.1:8443  -showcerts -CAfile bad_ca-chain.crt -verify_return_error -x509_strict
```

Still Happy Dance!!
```openssl
Server certificate
subject=C=US, ST=California, L=Temecula, O=ACME Inc, OU=BAD_ICA_IN_USE, CN=BAD Leaf Cert
issuer=C=US, ST=California, L=Temecula, O=ACME Inc, OU=BAD, CN=BAD ICA, emailAddress=bad@testing.io
---
No client certificate CA names sent
Peer signing digest: SHA256
Peer signature type: RSA-PSS
Server Temp Key: X25519, 253 bits
---
SSL handshake has read 5124 bytes and written 379 bytes
Verification: OK
---
New, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384
Server public key is 4096 bit
This TLS version forbids renegotiation.
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
```
4.6 Test with Erlang OTP 26 (in Erlang shell)
Command:
```erlang
ssl:start(), ssl:connect("127.0.0.1", 8443, [{verify, verify_peer}, {versions, ['tlsv1.3']}, {cacertfile, "bad_ca-chain.crt"}, {server_name_indication, disable}]).
```

Full Result:
```erlang
Erlang/OTP 26 [erts-14.2.5] [source] [64-bit] [smp:10:10] [ds:10:10:10] [async-threads:1] [jit] [dtrace]

Eshell V14.2.5 (press Ctrl+G to abort, type help(). for help)
1> ssl:start(), ssl:connect("127.0.0.1", 8443, [{verify, verify_peer}, {versions, ['tlsv1.3']}, {cacertfile, "bad_ca-chain.crt"}, {server_name_indication, disable}]).
=NOTICE REPORT==== 12-Sep-2024::11:10:04.724373 ===
TLS client: In state wait_cert_cr at ssl_handshake.erl:2117 generated CLIENT ALERT: Fatal - Unsupported Certificate

{error,{tls_alert,{unsupported_certificate,"TLS client: In state wait_cert_cr at ssl_handshake.erl:2117 generated CLIENT ALERT: Fatal - Unsupported Certificate\n"}}}
```

**Expected behavior**
Certificate chains with an ICA that contain a critical EKU extension are seen as having a supported extension.

**Affected versions**
Seen in both 25 and 26 major versions of Erlang OTP.

**Additional context**
See attached ZIP file for files produced in the test steps above.
[minimalRepo.zip](https://github.com/user-attachments/files/16983251/minimalRepo.zip)

### PR DESCRIPTION
**Pull request by IngelaAndin**
Should check CAs extended key usage conforms with key usage and therby handle possible critical extension.

Closes #8806

### CODE REVIEW

File: lib/public_key/src/pubkey_cert.erl [hunk 1]
```diff
@@ -1783,8 +1799,57 @@ is_digitally_sign_cert(Cert) ->
...
+is_compatible_purposes(_, []) ->
+    true;
+is_compatible_purposes(KeyUse, [?'id-kp-serverAuth'| Rest]) ->
+    (lists:member(digitalSignature, KeyUse) orelse
+     lists:member(keyAgreement, KeyUse)) andalso
+        is_compatible_purposes(KeyUse, Rest);
+is_compatible_purposes(KeyUse, [?'id-kp-clientAuth'| Rest]) ->
+    (lists:member(digitalSignature, KeyUse)
+     orelse
+       (lists:member(keyAgreement, KeyUse) or lists:member(keyEncipherment, KeyUse)))
```
Comment (dgud) [inline lib/public_key/src/pubkey_cert.erl:1821]:
```suggestion
       (lists:member(keyAgreement, KeyUse) orelse lists:member(keyEncipherment, KeyUse)))
```

File: lib/public_key/src/pubkey_cert.erl [hunk 2]
```diff
@@ -1783,8 +1799,57 @@ is_digitally_sign_cert(Cert) ->
...
+        andalso is_compatible_purposes(KeyUse, Rest);
+is_compatible_purposes(KeyUse, [?'id-kp-codeSigning'| Rest]) ->
+    lists:member(digitalSignature, KeyUse) andalso
+        is_compatible_purposes(KeyUse, Rest);
+is_compatible_purposes(KeyUse, [?'id-kp-emailProtection'| Rest]) ->
+    ((lists:member(digitalSignature, KeyUse) orelse
+      lists:member(nonRepudiation, KeyUse))
+     orelse
+       (lists:member(keyAgreement, KeyUse) or lists:member(keyEncipherment, KeyUse)))
```
Comment (dgud) [inline lib/public_key/src/pubkey_cert.erl:1830]:
```suggestion
       (lists:member(keyAgreement, KeyUse) orelse lists:member(keyEncipherment, KeyUse)))
```

**Review by dgud — COMMENTED** [review]
Maybe add a whitebox test of `compatible_ext_key_usage/2` so that everything is covered.

File: lib/public_key/src/pubkey_cert.erl [hunk 3]
```diff
@@ -1783,8 +1799,57 @@ is_digitally_sign_cert(Cert) ->
 	    lists:member(keyCertSign, KeyUse)
     end.
 
-missing_basic_constraints(Cert, SelfSigned, ValidationState, VerifyFun, UserState0,Len) ->
-    UserState = verify_fun(Cert, {bad_cert, missing_basic_constraint},
+compatible_ext_key_usage(_, undefined) ->
+    true;
+compatible_ext_key_usage(#'Extension'{extnValue = KeyUse}, #'Extension'{extnValue = Purposes}) ->
+    case lists:member(?anyExtendedKeyUsage, Purposes) of
```
Comment (u3s) [inline lib/public_key/src/pubkey_cert.erl:1805]: consider re-using `ext_keyusage_includes_any/1`

**Review by u3s — APPROVED** [review]

File: lib/public_key/src/pubkey_cert.erl (hunk 1 above)
Comment (IngelaAndin) [reply lib/public_key/src/pubkey_cert.erl:1821]: fixed

File: lib/public_key/src/pubkey_cert.erl (hunk 2 above)
Comment (IngelaAndin) [reply lib/public_key/src/pubkey_cert.erl:1830]: fixed

File: lib/public_key/src/pubkey_cert.erl (hunk 3 above)
Comment (IngelaAndin) [reply lib/public_key/src/pubkey_cert.erl:1805]: fixed

**Comment by IngelaAndin** [comment]
@dgud  Managed to extend basic test for better coverage. I dislike white-box test as they are more work to maintain. This test might not reflect common combination of key usage, but they work together theoretically and test the code better.

**Review by dgud — APPROVED** [review]

**Pull request closed without merging**